src/python/pants/engine/round_engine.py

```diff
@@ -43,6 +43,8 @@ def attempt(self, explain):
         with self._context.new_workunit(name=name, labels=[WorkUnitLabel.TASK], log_config=log_config):
           if explain:
             self._context.log.debug('Skipping execution of {} in explain mode'.format(name))
+          elif task.skip_execution:
```

Comment:
This subverts the product requirements. It seems like things should go one step further for `--skip` and the engine itself should be registering the option only for tasks that have no products or else no product consumers for the given run.

Reply:
I'm not sure I agree: I can imagine a situation where you don't want to create products (e.g., say you want to generate thrift but not proto, or something). Admittedly this is advanced usage, and you have to know what you're doing, but I don't think it's subversive.

Reply:
OK - I won't block on this since agreed it's advanced and could also be bolted on top of this change. That said, I can't see any cases where we couldn't follow an algorithm and ensure a user doesn't `--skip` a product needed as an intermediary; ie `./pants gen.thrift --skip //I-dont-know-why-but-im-specifying-target-with-thrift-and-proto-deps-here` could be allowed since the `gen` products were requested by the user directly by spelling the goal on the cli; ie: when a user says I want things and also I don't want them, that's ok. When they say instead `/pants compile gen.thrift --skip //I-dont-know-why-but-im-specifying-target-with-thrift-and-proto-deps-here` and compile deps on gen then the user just shot themselves in the foot and will get a compile error on missing thrift symbols.

Reply:
Well, when they say `./pants compile gen <targets>::` they get all the generators, and they might want to turn one of them off and only compile the results of generating some of them.

Reply:
A bit of a poor example, I mean this: `./pants compile --gen-thrift-skip //java/target/that/requires/thrift //scala/target/that/requires/proto`. That will blow up.

To be more transparent, the thought train was - in the new engine where only product dependencies activate tasks, forcing the skip of any interior node should always be fatal and the incorrect thing to do. In the current engine this is not the case because of the flat nature of goals (list of tasks) which are the nodes.